File: app/src/main/java/com/satsumasoftware/pokedex/framework/pokemon/MiniPokemon.java
package com.satsumasoftware.pokedex.framework.pokemon;

import android.content.Context;
import android.database.Cursor;
import android.os.Parcel;
import android.os.Parcelable;
import android.widget.ImageView;

import com.satsumasoftware.pokedex.db.PokemonDBHelper;
import com.satsumasoftware.pokedex.util.ActionUtils;
import com.satsumasoftware.pokedex.util.DataUtils;

public class MiniPokemon extends BasePokemon implements Parcelable {

    private int mAltDexId = -1, mAltDexNumber = -1;


    public MiniPokemon(Context context, int id) {
        PokemonDBHelper helper = new PokemonDBHelper(context);
        Cursor cursor = helper.getReadableDatabase().query(
                PokemonDBHelper.TABLE_NAME,
                DB_COLUMNS,
                PokemonDBHelper.COL_ID + "=? AND " + PokemonDBHelper.COL_FORM_IS_DEFAULT + "=?",
                new String[] {String.valueOf(id), "1"},
                null, null, null);
        cursor.moveToFirst();
        mId = id;
        mSpeciesId = cursor.getInt(cursor.getColumnIndex(PokemonDBHelper.COL_SPECIES_ID));
        mFormId = cursor.getInt(cursor.getColumnIndex(PokemonDBHelper.COL_FORM_ID));
        mName = cursor.getString(cursor.getColumnIndex(PokemonDBHelper.COL_NAME));
        mFormName = cursor.getString(cursor.getColumnIndex(PokemonDBHelper.COL_FORM_NAME));
        mFormPokemonName = cursor.getString(cursor.getColumnIndex(PokemonDBHelper.COL_FORM_POKEMON_NAME));
        mNationalNumber = cursor.getInt(cursor.getColumnIndex(PokemonDBHelper.COL_POKEDEX_NATIONAL));
        cursor.close();
    }

    public MiniPokemon(Cursor cursor) {  // TODO use this more often
        mId = cursor.getInt(cursor.getColumnIndex(PokemonDBHelper.COL_ID));
        mSpeciesId = cursor.getInt(cursor.getColumnIndex(PokemonDBHelper.COL_SPECIES_ID));
        mFormId = cursor.getInt(cursor.getColumnIndex(PokemonDBHelper.COL_FORM_ID));
        mName = cursor.getString(cursor.getColumnIndex(PokemonDBHelper.COL_NAME));
        mFormName = cursor.getString(cursor.getColumnIndex(PokemonDBHelper.COL_FORM_NAME));
        mFormPokemonName = cursor.getString(cursor.getColumnIndex(PokemonDBHelper.COL_FORM_POKEMON_NAME));
        mNationalNumber = cursor.getInt(cursor.getColumnIndex(PokemonDBHelper.COL_POKEDEX_NATIONAL));
    }

    public MiniPokemon(int id, int speciesId, int formId, String name,
                       String formName, String formPokemonName, int nationalDexNumber) {
        mId = id;
        mSpeciesId = speciesId;
        mFormId = formId;
        mName = name;
        mFormName = formName;
        mFormPokemonName = formPokemonName;
        mNationalNumber = nationalDexNumber;
    }

    @Deprecated
    public static MiniPokemon createFromSpecies(Context context, int speciesId, boolean isFormMega) {
        MiniPokemon miniPokemon = null;
        int isMegaAsInt = (isFormMega ? 1 : 0);
        PokemonDBHelper pokemonDBHelper = new PokemonDBHelper(context);
        Cursor cursor = pokemonDBHelper.getReadableDatabase().query(
                PokemonDBHelper.TABLE_NAME,
                DB_COLUMNS,
                PokemonDBHelper.COL_SPECIES_ID + "=? AND " + PokemonDBHelper.COL_FORM_IS_MEGA + "=?",
                new String[] {String.valueOf(speciesId), String.valueOf(isMegaAsInt)},
                null, null, null);
        cursor.moveToFirst();
        while (!cursor.isAfterLast()) {
            miniPokemon = new MiniPokemon(cursor);  // TODO did I mean to use moveToNext() here?
        }
        cursor.close();
        return miniPokemon;

    }

    public Pokemon toPokemon(Context context) {
        return new Pokemon(context, mId, mSpeciesId, mFormId, mName, mFormName, mFormPokemonName,
                mNationalNumber);
    }

    public void addAlternatePokedexInfo(int pokedexId, int pokedexNumber) {
        mAltDexId = pokedexId;
        mAltDexNumber = pokedexNumber;
    }

    public boolean hasAddedAltDexInfo() {
        return mAltDexNumber != DataUtils.NULL_INT;
    }

    public int getAlternateDexId() {
        return mAltDexId;
    }

    public int getAlterateDexNumber() {
        return mAltDexNumber;
    }


    public void setPokemonImage(ImageView imageView) {
        ActionUtils.setPokemonImage(this, imageView);
    }


    protected MiniPokemon(Parcel in) {
        mId = in.readInt();
        mSpeciesId = in.readInt();
        mFormId = in.readInt();
        mNationalNumber = in.readInt();
        mAltDexId = in.readInt();
        mAltDexNumber = in.readInt();

        mName = in.readString();
        mFormName = in.readString();
        mFormPokemonName = in.readString();
    }


    public static final Parcelable.Creator<MiniPokemon> CREATOR = new Parcelable.Creator<MiniPokemon>() {
        @Override
        public MiniPokemon createFromParcel(Parcel in) {
            return new MiniPokemon(in);
        }

        @Override
        public MiniPokemon[] newArray(int size) {
            return new MiniPokemon[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(mId);
        dest.writeInt(mSpeciesId);
        dest.writeInt(mFormId);
        dest.writeInt(mNationalNumber);
        dest.writeInt(mAltDexId);
        dest.writeInt(mAltDexNumber);

        dest.writeString(mName);
        dest.writeString(mFormName);
        dest.writeString(mFormPokemonName);
    }

}

File: app/src/main/java/com/satsumasoftware/pokedex/framework/pokemon/MiniPokemon.kt
package com.satsumasoftware.pokedex.framework.pokemon

import android.content.Context
import android.database.Cursor
import java.util.*
import android.os.Parcel
import android.os.Parcelable
import android.widget.ImageView
import com.satsumasoftware.pokedex.db.PokemonDBHelper
import com.satsumasoftware.pokedex.util.ActionUtils

class MiniPokemon(id: Int, speciesId: Int, formId: Int, name: String, formName: String,
                  formAndPokemonName: String?, nationalNumber: Int) :
        BasePokemon(id, speciesId, formId, name, formName, formAndPokemonName, nationalNumber),
        Parcelable {

    constructor(cursor: Cursor) : this(
            cursor.getInt(cursor.getColumnIndex(PokemonDBHelper.COL_ID)),
            cursor.getInt(cursor.getColumnIndex(PokemonDBHelper.COL_SPECIES_ID)),
            cursor.getInt(cursor.getColumnIndex(PokemonDBHelper.COL_FORM_ID)),
            cursor.getString(cursor.getColumnIndex(PokemonDBHelper.COL_NAME)),
            cursor.getString(cursor.getColumnIndex(PokemonDBHelper.COL_FORM_NAME)),
            cursor.getString(cursor.getColumnIndex(PokemonDBHelper.COL_FORM_POKEMON_NAME)),
            cursor.getInt(cursor.getColumnIndex(PokemonDBHelper.COL_POKEDEX_NATIONAL)))

    constructor(source: Parcel) : this(
            source.readInt(),
            source.readInt(),
            source.readInt(),
            source.readString(),
            source.readString(),
            source.readString(),
            source.readInt())

    fun toPokemon(context: Context) = Pokemon(context,
            id,
            speciesId,
            formId,
            name,
            formName,
            formAndPokemonName,
            nationalDexNumber)

    fun setPokemonImage(imageView: ImageView) = ActionUtils.setPokemonImage(this, imageView)

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel?, flags: Int) {
        dest?.writeInt(id)
        dest?.writeInt(speciesId)
        dest?.writeInt(formId)
        dest?.writeString(name)
        dest?.writeString(formName)
        dest?.writeString(formAndPokemonName)
        dest?.writeInt(nationalDexNumber)
    }

    companion object {

        @JvmField val CREATOR: Parcelable.Creator<MiniPokemon> = object : Parcelable.Creator<MiniPokemon> {
            override fun createFromParcel(source: Parcel): MiniPokemon = MiniPokemon(source)
            override fun newArray(size: Int): Array<MiniPokemon?> = arrayOfNulls(size)
        }

        @JvmStatic
        fun create(context: Context, id: Int): MiniPokemon {
            val helper = PokemonDBHelper.getInstance(context)
            val cursor = helper.readableDatabase.query(
                    PokemonDBHelper.TABLE_NAME,
                    BasePokemon.DB_COLUMNS,
                    "${PokemonDBHelper.COL_ID}=? AND ${PokemonDBHelper.COL_FORM_IS_DEFAULT}=?",
                    arrayOf(id.toString(), 1.toString()),
                    null, null, null)
            cursor.moveToFirst()
            val miniPokemon = MiniPokemon(cursor)
            cursor.close()
            return miniPokemon
        }

        @JvmStatic
        fun createFromSpecies(context: Context, speciesId: Int, isFormMega: Boolean): MiniPokemon {
            val isMegaAsInt = if (isFormMega) 1 else 0
            val pokemonDBHelper = PokemonDBHelper.getInstance(context)
            val cursor = pokemonDBHelper.readableDatabase.query(
                    PokemonDBHelper.TABLE_NAME,
                    BasePokemon.DB_COLUMNS,
                    "${PokemonDBHelper.COL_SPECIES_ID}=? AND ${PokemonDBHelper.COL_FORM_IS_MEGA}=?",
                    arrayOf(speciesId.toString(), isMegaAsInt.toString()),
                    null, null, null)
            cursor.moveToFirst()
            val miniPokemon = MiniPokemon(cursor)
            cursor.close()
            return miniPokemon
        }
    }

}
